#########################################################################################################
#  Groups Statistics
#
#  The inter-event distance for a uniform distribution of random events can be approximated with a 
#  Poisson distribution; this is equivalent to the waiting time distribution. 
#  The probability density of an inter-event distance D in a chromosome is:
#               f(D) = λe^(-λD), 
#                                   where λ = N⁄L is the total rate of mutations, 
#                                             N   is the total number of mutations, and 
#                                             L   is the total number of base pairs in the chromosome.  
#
#   If distances are coarse grained into bins of size D_bin, then P_low = 1⁄(N∙D_bin) 
#   is the smallest probability we can detect (given the binning resolution)
#   
#   Given F(X) = ∫_{0}^{X} f(x)dx, then 
#   (a) The probability of an interval D <= D' : F(D) = 1 - e^(-λD')
#   (b) The probability of an interval D > D' (survival function) : S(D) = e^(-λD')
#
#   The probability associated with each bin [di, di + D_bin] will be given as:
#   P(di) = F(di + D_bin) - F(di) = [1 - e^(-λ(di + D_bin)] - [1 - e^(-λdi)]
#   which is used to build a histogram of expected counts for each interval bin range.
#
####
#
#   If we define a *tuple* as a set of consecutive mutations that are closer than a distance D_grp.  
#   The probability of inter-event interval D<D_grp :  P_ = 1 - e^(-λD_grp) 
#   The probability of D>D_grp :                       P+ = e^(-λD_grp ) 
#
#    Considering singleton mutations as 1-tuples, then all groups are separated by intervals 
#    D > D_grp, thus the number of groups in the chromosome is estimated as N_grp = N∙P+, 
#    with N- = (1 + N∙P-) the number of events in groups.
#
#    The probability distribution density of group sizes can be written as 
#     P_grp (s) = (P+)^2(P-)^(s-1), where s is the group size. 
#
#    Therefore the expected number of groups of size s is   N_grp (s) = N∙P_grp(s) 
#    and the frequency of groups of size s is given as      
#                          P_s = (N_grp(s))⁄N_grp = P_+ (P_-)^(s-1) 
#    which is equivalent to a binomial term of first order
#
#########################################################################################################
#
#  Calculated Statistics: 
#    (a) Distribution of inter-event distances, 
#    (b) Distribution of group sizes,
#    (c) Histogram of events in interval size bins as a function of N_SNV 
#        (done for low interval sizes)
#   
#   (a) and (b) are generated by selecting 3 samples (min, med and max in Chi-squared).
#
#########################################################################################################
rm(list = ls())

library(fitdistrplus)
library(goftest)
library(rstudioapi) 
library(ggplot2)
library(data.table);
library(grid)
library(gridExtra)
library(lattice)
library(scales)
library(readr)
library(RCircos)

setwd(dirname(getActiveDocumentContext()$path)) 

load("clusters.RData") 

if (sim_flag) {
  
  sim.pchi    <- pchi(sim.dist.tbl)
  sim.penrich <- penrich(sim.dist.tbl)
  #sim.p       <- p.goft(sim.eids.tbl)
  
}

########################################################################################################################

if (cgi_flag) {
  
  cgi.pchi    <- pchi(cgi.dist.tbl)
  cgi.penrich <- penrich(cgi.dist.tbl)
  #cgi.p       <- p.goft(cgi.eids.tbl)

}
########################################################################################################################

if (ca1_flag) {
  
  ca1.pchi    <- pchi(ca1.dist.tbl)
  ca1.penrich <- penrich(ca1.dist.tbl)
  #ca1.p       <- p.goft(ca1.eids.tbl)
  
}

#########################################################################################################

low.D.freq <- ggplot()
low.S.freq <- ggplot()

if (ca1_flag) {
  
  ca1.plots <- plots(DT       = ca1.pchi, 
                     dist.tbl = ca1.dist.tbl, 
                     xlim1    = c(8, 3e4),  
                     ylim1    = c(1e-5, .9), 
                     xbreaks1 = c(1,10,100,100,1000,10000,100000), 
                     ybreaks1 = c(1e-6,1e-4,1e-2,1),
                     grps.tbl = ca1.grps.tbl, 
                     xlim2    = c(0.8, 14),   
                     ylim2    = c(1e-5, 1.5),
                     xbreaks2 = seq(0, 30, 2), 
                     ybreaks2 = c(1e-4,1e-2,1))
  
  ca1.low.D.freq <- d.low.freq(DT      = ca1.penrich, 
                               DTref   = ca1.penrich,
                               Dmax    = Dmax, 
                               xlim    = c(350, 900000), 
                               ylim    = c(1, 100000), 
                               xbreaks = c(100,1000,10000,100000,1000000), 
                               ybreaks = c(1,100,10000,1000000),
                               levs    = levs_D, 
                               cols    = cols_D, 
                               nams    = nams_D)
  
  ca1.low.S.freq <- s.low.freq(DT      = ca1.grps.tbl, 
                               DTref   = ca1.grps.tbl,
                               Smax    = Smax,  
                               xlim    = c(350, 600000), 
                               ylim    = c(.5, 60000), 
                               xbreaks = c(100,1000,10000,100000,1000000), 
                               ybreaks = c(.1, 1,100,10000,1000000),
                               levs    = levs_S, 
                               cols    = cols_S, 
                               nams    = nams_S)
  
  aux <- ca1.penrich[D > levs_D[1], .(SID, N_SNV, N_D)]
  aux <- aux[, N_D := sum(N_D), by = .(SID)]
  aux <- unique(aux[, .(SID, N_SNV, N_D)])
  low.D.freq <- low.D.freq + 
    geom_point( data = ca1.penrich[D == levs_D[1]],
                aes(x      = N_SNV, 
                    y      = N_D,
                    colour = "ca1.1"),
                shape  = 20,  
                size   = 1
    ) +
    geom_point( data = aux,
                aes(x      = N_SNV, 
                    y      = N_D,
                    colour = "ca1.2"),
                shape  = 20,  
                size   = 1) 
  
  aux <- ca1.grps.tbl[group_Size > 1, .(SID, N_SNV, N_S)]
  aux <- aux[, N_S := sum(N_S), by = .(SID)]
  aux <- unique(aux[, .(SID, N_SNV, N_S)])
  low.S.freq <- low.S.freq + 
    geom_point( data = ca1.grps.tbl[group_Size == 1],
                aes(x      = N_SNV, 
                    y      = N_S,
                    colour = "ca1.1"),
                shape  = 20,  
                size   = 1
    ) +
    geom_point( data = aux,
                aes(x      = N_SNV, 
                    y      = N_S,
                    colour = "ca1.2"),
                shape  = 20,  
                size   = 1)
  rm(aux)
  
  ca1.dist.freq <- ca1.plots[[1]]
  ca1.grps.freq <- ca1.plots[[2]]
  
}

########################################################################################################################

if (cgi_flag) {
  
  cgi.plots <- plots(DT       = cgi.pchi, 
                     dist.tbl = cgi.dist.tbl, 
                     xlim1    = c(8, 3e4),  
                     ylim1    = c(1e-5, .9), 
                     xbreaks1 = c(1,10,100,100,1000,10000,100000), 
                     ybreaks1 = c(1e-6,1e-4,1e-2,1),
                     grps.tbl = cgi.grps.tbl, 
                     xlim2    = c(0.8, 14),   
                     ylim2    = c(1e-5, 1.5),
                     xbreaks2 = seq(0, 30, 2), 
                     ybreaks2 = c(1e-4,1e-2,1))
  
  cgi.low.D.freq <- d.low.freq(DT      = cgi.penrich, 
                               DTref   = ca1.penrich,
                               Dmax    = Dmax,
                               xlim    = c(350, 900000), 
                               ylim    = c(1, 100000), 
                               xbreaks = c(100,1000,10000,100000,1000000), 
                               ybreaks = c(1,100,10000,1000000),
                               levs    = levs_D, 
                               cols    = cols_D, 
                               nams    = nams_D)
  
  cgi.low.S.freq <- s.low.freq(DT      = cgi.grps.tbl, 
                               DTref   = ca1.grps.tbl,
                               Smax    = Smax,
                               xlim    = c(350, 600000), 
                               ylim    = c(.5, 60000), 
                               xbreaks = c(100,1000,10000,100000,1000000), 
                               ybreaks = c(.1, 1,100,10000,1000000),
                               levs    = levs_S, 
                               cols    = cols_S, 
                               nams    = nams_S)
  
  aux <- cgi.penrich[D > levs_D[1], .(SID, N_SNV, N_D)]
  aux <- aux[, N_D := sum(N_D), by = .(SID)]
  aux <- unique(aux[, .(SID, N_SNV, N_D)])
  low.D.freq <- low.D.freq + 
    geom_point( data = cgi.penrich[D == levs_D[1]],
                aes(x      = N_SNV, 
                    y      = N_D,
                    colour = "cgi.1"),
                shape  = 20,  
                size   = 1
    ) +
    geom_point( data = aux,
                aes(x      = N_SNV, 
                    y      = N_D,
                    colour = "cgi.2"),
                shape  = 18,  
                size   = 1) 
  
  aux <- cgi.grps.tbl[group_Size > 1, .(SID, N_SNV, N_S)]
  aux <- aux[, N_S := sum(N_S), by = .(SID)]
  aux <- unique(aux[, .(SID, N_SNV, N_S)])
  low.S.freq <- low.S.freq + 
    geom_point( data = cgi.grps.tbl[group_Size == 1],
                aes(x      = N_SNV, 
                    y      = N_S,
                    colour = "cgi.1"),
                shape  = 20,  
                size   = 1
    ) +
    geom_point( data = aux,
                aes(x      = N_SNV, 
                    y      = N_S,
                    colour = "cgi.2"),
                shape  = 18,  
                size   = 1)
  rm(aux)
  
  cgi.dist.freq <- cgi.plots[[1]]
  cgi.grps.freq <- cgi.plots[[2]]
  
}

########################################################################################################################

if (sim_flag) {
  
  sim.plots <- plots(DT       = sim.pchi, 
                     dist.tbl = sim.dist.tbl, 
                     xlim1    = c(8, 3e4),  
                     ylim1    = c(1e-5, .9), 
                     xbreaks1 = c(1,10,100,100,1000,10000,100000), 
                     ybreaks1 = c(1e-6,1e-4,1e-2,1),
                     grps.tbl = sim.grps.tbl, 
                     xlim2    = c(0.8, 14),   
                     ylim2    = c(1e-5, 1.5),
                     xbreaks2 = seq(0, 30, 1), 
                     ybreaks2 = c(1e-4,1e-2,1))
  
  sim.low.D.freq <- d.low.freq(DT      = sim.penrich, 
                               DTref   = ca1.penrich,
                               Dmax    = Dmax,
                               xlim    = c(350, 900000), 
                               ylim    = c(1, 100000), 
                               xbreaks = c(100,1000,10000,100000,1000000), 
                               ybreaks = c(1,100,10000,1000000),
                               levs    = levs_D, 
                               cols    = cols_D, 
                               nams    = nams_D)
  
  sim.low.S.freq <- s.low.freq(DT      = sim.grps.tbl, 
                               DTref   = ca1.grps.tbl,
                               Smax    = Smax,
                               xlim    = c(350, 600000), 
                               ylim    = c(.5, 60000), 
                               xbreaks = c(100,1000,10000,100000,1000000), 
                               ybreaks = c(.1, 1,100,10000,1000000),
                               levs    = levs_S, 
                               cols    = cols_S, 
                               nams    = nams_S)
  
  aux <- sim.penrich[D > levs_D[1], .(SID, N_SNV, N_D)]
  aux <- aux[, N_D := sum(N_D), by = .(SID)]
  aux <- unique(aux[, .(SID, N_SNV, N_D)])
  low.D.freq <- low.D.freq + 
    geom_point( data = sim.penrich[D == levs_D[1]],
                aes(x      = N_SNV, 
                    y      = N_D,
                    colour = "sim.1"),
                shape  = 20,  
                size   = 1
    ) +
    geom_point( data = aux,
                aes(x      = N_SNV, 
                    y      = N_D,
                    colour = "sim.2"),
                shape  = 18,  
                size   = 1) 
  
  aux <- sim.grps.tbl[group_Size > 1, .(SID, N_SNV, N_S)]
  aux <- aux[, N_S := sum(N_S), by = .(SID)]
  aux <- unique(aux[, .(SID, N_SNV, N_S)])
  low.S.freq <- low.S.freq + 
    geom_point( data = sim.grps.tbl[group_Size == 1],
                aes(x      = N_SNV, 
                    y      = N_S,
                    colour = "sim.1"),
                shape  = 20,  
                size   = 1
    ) +
    geom_point( data = aux,
                aes(x      = N_SNV, 
                    y      = N_S,
                    colour = "sim.2"),
                shape  = 18,  
                size   = 1)
  rm(aux)
  
  sim.dist.freq <- sim.plots[[1]]
  sim.grps.freq <- sim.plots[[2]]
  
}

########################################################################################################################

low.D.freq <- low.D.freq + 
  geom_line( data = ca1.penrich[D == levs_D[1]],
             aes(x      = N_SNV, 
                 y      = E_D,
                 colour = "sim.1"),
             linetype = "dashed",  
             size   = .5
  ) +
  geom_line( data = ca1.penrich[D == levs_D[1]],
             aes(x      = N_SNV, 
                 y      = N_SNV - E_D,
                 colour = "sim.2"),
             linetype = "dashed",  
             size   = .5
  ) +
  coord_cartesian(xlim    = c(300, 900000),
                  ylim    = c(.5, 400000),
                  expand  = FALSE
  ) +
  scale_x_log10(breaks       = logbreaks,
                minor_breaks = logminor_breaks,
                labels       = comma
  ) +
  scale_y_log10(breaks       = c(1,100,10000,1000000),
                minor_breaks = logminor_breaks,
                labels       = comma
  ) +
  scale_colour_manual(values = cols.sub,
                      breaks = c("ca1.1", 
                                 "ca1.2",
                                 "cgi.1",
                                 "cgi.2"),
                      labels = c("ca1.1" = "Cancer (PCAWG) - D < 15Kb", 
                                 "ca1.2" = "Cancer (PCAWG) - D > 15Kb",
                                 "cgi.1" = "Normal (CGI) - D < 15Kb",
                                 "cgi.2" = "Normal (CGI) - D > 15Kb"),
                      drop   = TRUE
  ) + 
  theme(text         = element_text(size   = 24, 
                                    colour = "black", 
                                    face   = "bold"),
        axis.text    = element_text(size   = rel(0.75)),
        axis.title.x = element_text(size   = rel(1),
                                    margin = margin(1,0,0,0, unit = "lines")),
        axis.text.x  = element_text(angle  = 0, 
                                    hjust  = 0.5),
        axis.title.y = element_text(size   = rel(1),
                                    margin = margin(0,1,0,0, unit = "lines")),
        axis.text.y  = element_text(angle  = 90, 
                                    hjust  = 0.5),
        plot.margin  = unit(c(1, 1, 1, 1), "lines"), #(up, right, down, left)
        plot.title   = element_text(size   = rel(1), 
                                    hjust  = 0,
                                    margin = margin(0,0,1,0, unit = "lines")),
        legend.title          = element_blank(),
        legend.position       = c(0.99, 0.01),
        legend.justification  = c(1, 0),
        legend.text           = element_text(size = 10),
        legend.box.just       = "right", 
        legend.box.background = element_rect()
  ) + 
  guides(colour = guide_legend(override.aes = list( size     = 5,
                                                    linetype = 0),
                               drop   = TRUE)
  ) +
  labs( x = "Total Number of SNVs",
        y = "Number of Tuples"
  ) 

print(low.D.freq)

########################################################################################################################

low.S.freq  <- low.S.freq + 
  geom_line( data = ca1.grps.tbl[group_Size == 1],
             aes(x      = N_SNV, 
                 y      = N_siz,
                 colour = "sim.1"),
             linetype = "dashed",  
             size   = .5
  ) +
  geom_line( data = ca1.grps.tbl[group_Size == 1],
             aes(x      = N_SNV, 
                 y      = N_grps - N_siz,
                 colour = "sim.2"),
             linetype = "dashed",  
             size   = .5
  ) +
  coord_cartesian(xlim    = c(300, 900000),
                  ylim    = c(1, 90000),
                  expand  = FALSE
  ) +
  scale_x_log10(breaks       = logbreaks,
                minor_breaks = logminor_breaks,
                labels       = comma
  ) +
  scale_y_log10(breaks       = logbreaks,
                minor_breaks = logminor_breaks,
                labels       = comma
  ) +
  scale_colour_manual(values = cols.sub,
                      breaks = c("ca1.1", 
                                 "ca1.2",
                                 "cgi.1",
                                 "cgi.2"),
                      labels = c("ca1.1" = "Cancer (PCAWG) - Singletons",
                                 "ca1.2" = "Cancer (PCAWG) - Tuples",
                                 "cgi.1" = "Normal (CGI) - Singletons",
                                 "cgi.2" = "Normal (CGI) - Tuples"),
                      drop   = TRUE
  ) + 
  theme(text         = element_text(size   = 24, 
                                    colour = "black", 
                                    face   = "bold"),
        axis.text    = element_text(size   = rel(0.75)),
        axis.title.x = element_text(size   = rel(1),
                                    margin = margin(1,0,0,0, unit = "lines")),
        axis.text.x  = element_text(angle  = 0, 
                                    hjust  = 0.5),
        axis.title.y = element_text(size   = rel(1),
                                    margin = margin(0,1,0,0, unit = "lines")),
        axis.text.y  = element_text(angle  = 90, 
                                    hjust  = 0.5),
        plot.margin  = unit(c(1, 1, 1, 1), "lines"), #(up, right, down, left)
        plot.title   = element_text(size   = rel(1), 
                                    hjust  = 0,
                                    margin = margin(0,0,1,0, unit = "lines")),
        legend.title          = element_blank(),
        legend.position       = c(0.99, 0.01),
        legend.justification  = c(1, 0),
        legend.text           = element_text(size = 10),
        legend.box.just       = "right", 
        legend.box.background = element_rect()
  ) + 
  guides(colour = guide_legend(override.aes = list( size     = 5,
                                                    linetype = 0))
  ) +
  labs( x = "Total Number of SNVs",
        y = "Number of Tuples"
  ) 

print(low.S.freq)

########################################################################################################################

num <- 0
if (sim_flag) {
  num <- num + 1
} 
if (cgi_flag) {
  num <- num + 1
}     
if (ca1_flag) {
  num <- num + 1
}   

if (num>0) {
  
  fileout <- paste("fig_tuples.0.tiff", sep="")
  tiff(file=fileout, width = 30 , height = 15, units = "in", res = 100, bg = "white", type = "cairo")
  
  vplayout <- function(x, y) viewport(layout.pos.row = x, layout.pos.col = y)
  grid.newpage()
  pushViewport(viewport(layout = grid.layout(2, num)))
  
  n <- 1
  if (sim_flag) {
    print(sim.dist.freq  + ggtitle("Simulated Data"), vp = vplayout(1, n))
    print(sim.grps.freq  + ggtitle("Simulated Data"), vp = vplayout(2, n))
    n <- n + 1
  }
  if (cgi_flag) {
    print(cgi.dist.freq  + ggtitle("Normal (CGI)"),   vp = vplayout(1, n))
    print(cgi.grps.freq  + ggtitle("Normal (CGI)"),   vp = vplayout(2, n))
    n <- n + 1
  } 
  if (ca1_flag) {
    print(ca1.dist.freq  + ggtitle("Cancer (PCAWG)"), vp = vplayout(1, n))
    print(ca1.grps.freq  + ggtitle("Cancer (PCAWG)"), vp = vplayout(2, n))
    n <- n + 1
  } 
  dev.off()
  
  ########################################################################################################################
  
  fileout <- paste("fig_tuples.1.tiff", sep="")
  tiff(file=fileout, width = 30 , height = 15, units = "in", res = 100, bg = "white", type = "cairo")
  
  vplayout <- function(x, y) viewport(layout.pos.row = x, layout.pos.col = y)
  grid.newpage()
  pushViewport(viewport(layout = grid.layout(2, num)))
  
  n <- 1
  if (sim_flag) {
    print(sim.low.D.freq + ggtitle("Simulated Data"), vp = vplayout(1, n))
    print(sim.low.S.freq + ggtitle("Simulated Data"), vp = vplayout(2, n))
    n <- n + 1
  }
  if (cgi_flag) {
    print(cgi.low.D.freq + ggtitle("Normal (CGI)"),   vp = vplayout(1, n))
    print(cgi.low.S.freq + ggtitle("Normal (CGI)"),   vp = vplayout(2, n))
    n <- n + 1
  } 
  if (ca1_flag) {
    print(ca1.low.D.freq + ggtitle("Cancer (PCAWG)"), vp = vplayout(1, n))
    print(ca1.low.S.freq + ggtitle("Cancer (PCAWG)"), vp = vplayout(2, n))
    n <- n + 1
  } 
  dev.off()
  
}

########################################################################################################################

leg.plt      <- sim.low.D.freq + 
  theme(legend.title          = element_text(face = "bold",
                                             hjust = 0.5),
        legend.title.align    = 0.5,
        legend.position       = c(.5, .5),
        legend.justification  = c(.5, .5),
        legend.text           = element_text(size = 18),
        legend.box.just       = "center", 
        legend.box.background = element_rect(),
        legend.key.height     = unit(3, "line"),
        legend.key.width      = unit(3, "line")
  ) +
  labs( color = "Inter-SNV distance range [kb]\n       "
  ) +
  guides(color = guide_legend(title.hjust  = 0.5,
                              title.vjust  = 0.0,
                              ncol         = 2,
                              override.aes = list(size = 8)))

mylegend <- get_legend(leg.plt)

fileout <- paste("fig_tuples.1a.tiff", sep="")
tiff(file=fileout, width = 18 , height = 12, units = "in", res = 100, bg = "white", type = "cairo")

vplayout <- function(x, y) viewport(layout.pos.row = x, layout.pos.col = y)
grid.newpage()
grid.arrange(sim.low.D.freq + ggtitle("(A) Simulated Samples") + theme(legend.position = "none"), 
             mylegend,
             cgi.low.D.freq + ggtitle("(B) Normal Samples (CGI)") + theme(legend.position = "none"), 
             ca1.low.D.freq + ggtitle("(C) Cancer Samples (PCAWG)") + theme(legend.position = "none"), 
             layout_matrix = rbind(c(1, 3, 2), c(4, 5, 2)))
dev.off()

########################################################################################################################

leg.plt      <- sim.low.S.freq + 
  theme(legend.title          = element_text(face = "bold",
                                             hjust = 0.5),
        legend.title.align    = 0.5,
        legend.position       = c(.5, .5),
        legend.justification  = c(.5, .5),
        legend.text           = element_text(size = 18),
        legend.box.just       = "center", 
        legend.box.background = element_rect(),
        legend.key.height     = unit(3, "line"),
        legend.key.width      = unit(3, "line")
  ) +
  labs( color = "n-Tuple size \n       "
  ) +
  guides(color = guide_legend(title.hjust  = 0.5,
                              title.vjust  = 0.0,
                              ncol         = 2,
                              override.aes = list(size = 8)))

mylegend <- get_legend(leg.plt)

fileout <- paste("fig_tuples.1b.tiff", sep="")
tiff(file=fileout, width = 18 , height = 12, units = "in", res = 100, bg = "white", type = "cairo")

vplayout <- function(x, y) viewport(layout.pos.row = x, layout.pos.col = y)
grid.newpage()
grid.arrange(sim.low.S.freq + ggtitle("(A) Simulated Samples") + theme(legend.position = "none"), 
             mylegend,
             cgi.low.S.freq + ggtitle("(B) Normal Samples (CGI)") + theme(legend.position = "none"), 
             ca1.low.S.freq + ggtitle("(C) Cancer Samples (PCAWG)") + theme(legend.position = "none"), 
             layout_matrix = rbind(c(1, 3, 2), c(4, 5, 2)))
dev.off()

########################################################################################################################

print(low.S.freq)

fileout <- paste("fig_tuples.2.tiff", sep="")
tiff(file=fileout, width = 15 , height = 7.5, units = "in", res = 100, bg = "white", type = "cairo")

vplayout <- function(x, y) viewport(layout.pos.row = x, layout.pos.col = y)
grid.newpage()
pushViewport(viewport(layout = grid.layout(1, 2)))
print(low.D.freq, vp = vplayout(1, 1))  
print(low.S.freq, vp = vplayout(1, 2))
dev.off()

#########################################################################################################

fileout <- paste("fig_tuples.2a.tiff", sep="")
tiff(file=fileout, width = 7.5 , height = 7.5, units = "in", res = 300, bg = "white", type = "cairo")
print(low.D.freq)
dev.off()

fileout <- paste("fig_tuples.2b.tiff", sep="")
tiff(file=fileout, width = 7.5 , height = 7.5, units = "in", res = 300, bg = "white", type = "cairo")
print(low.S.freq)
dev.off()

if (sim_flag) {
  fileout <- paste("fig_tuples.0Aa.tiff", sep="")
  #tiff(file=fileout, width = 7.5 , height = 7.5, units = "in", res = 300, bg = "white", type = "cairo")
  tiff(file=fileout, width = 4500 , height = 4500, units = "px", type="cairo", res = 300)
  print(sim.dist.freq + ggtitle("Simulated Data"))
  dev.off()
  
  fileout <- paste("fig_tuples.0Ab.tiff", sep="")
  #tiff(file=fileout, width = 7.5 , height = 7.5, units = "in", res = 300, bg = "white", type = "cairo")
  tiff(file=fileout, width = 4500 , height = 4500, units = "px", type="cairo", res = 300)
  print(sim.grps.freq + ggtitle("Simulated Data"))
  dev.off()
  
  fileout <- paste("fig_tuples.1Aa.tiff", sep="")
  #tiff(file=fileout, width = 7.5 , height = 7.5, units = "in", res = 300, bg = "white", type = "cairo")
  tiff(file=fileout, width = 4500 , height = 4500, units = "px", type="cairo", res = 300)
  print(sim.low.D.freq + ggtitle("Simulated Data"))
  dev.off()
  
  fileout <- paste("fig_tuples.1Ab.tiff", sep="")
  #tiff(file=fileout, width = 7.5 , height = 7.5, units = "in", res = 300, bg = "white", type = "cairo")
  tiff(file=fileout, width = 4500 , height = 4500, units = "px", type="cairo", res = 300)
  print(sim.low.S.freq + ggtitle("Simulated Data"))
  dev.off()
}

if (cgi_flag) {
  fileout <- paste("fig_tuples.0Ba.tiff", sep="")
  #tiff(file=fileout, width = 7.5 , height = 7.5, units = "in", res = 300, bg = "white", type = "cairo")
  tiff(file=fileout, width = 4500 , height = 4500, units = "px", type="cairo", res = 300)
  print(cgi.dist.freq + ggtitle("Normal (CGI)"))
  dev.off()
  
  fileout <- paste("fig_tuples.0Bb.tiff", sep="")
  #tiff(file=fileout, width = 7.5 , height = 7.5, units = "in", res = 300, bg = "white", type = "cairo")
  tiff(file=fileout, width = 4500 , height = 4500, units = "px", type="cairo", res = 300)
  print(cgi.grps.freq + ggtitle("Normal (CGI)"))
  dev.off()
  
  fileout <- paste("fig_tuples.1Ba.tiff", sep="")
  #tiff(file=fileout, width = 7.5 , height = 7.5, units = "in", res = 300, bg = "white", type = "cairo")
  tiff(file=fileout, width = 4500 , height = 4500, units = "px", type="cairo", res = 300)
  print(cgi.low.D.freq + ggtitle("Normal (CGI)"))
  dev.off()
  
  fileout <- paste("fig_tuples.1Bb.tiff", sep="")
  #tiff(file=fileout, width = 7.5 , height = 7.5, units = "in", res = 300, bg = "white", type = "cairo")
  tiff(file=fileout, width = 4500 , height = 4500, units = "px", type="cairo", res = 300)
  print(cgi.low.S.freq + ggtitle("Normal (CGI)"))
  dev.off()
}

if (ca1_flag) {
  
  fileout <- paste("fig_tuples.0Ca.tiff", sep="")
  #tiff(file=fileout, width = 7.5 , height = 7.5, units = "in", res = 300, bg = "white", type = "cairo")
  tiff(file=fileout, width = 4500 , height = 4500, units = "px", type="cairo", res = 300)
  print(ca1.dist.freq + ggtitle("Cancer (PCAWG)"))
  dev.off()
  
  fileout <- paste("fig_tuples.0Cb.tiff", sep="")
  #tiff(file=fileout, width = 7.5 , height = 7.5, units = "in", res = 300, bg = "white", type = "cairo")
  tiff(file=fileout, width = 4500 , height = 4500, units = "px", type="cairo", res = 300)
  print(ca1.grps.freq + ggtitle("Cancer (PCAWG)"))
  dev.off()
  
  fileout <- paste("fig_tuples.1Ca.tiff", sep="")
  #tiff(file=fileout, width = 7.5 , height = 7.5, units = "in", res = 300, bg = "white", type = "cairo")
  tiff(file=fileout, width = 4500 , height = 4500, units = "px", type="cairo", res = 300)
  print(ca1.low.D.freq + ggtitle("Cancer (PCAWG)"))
  dev.off()
  
  fileout <- paste("fig_tuples.1Cb.tiff", sep="")
  #tiff(file=fileout, width = 7.5 , height = 7.5, units = "in", res = 300, bg = "white", type = "cairo")
  tiff(file=fileout, width = 4500 , height = 4500, units = "px", type="cairo", res = 300)
  print(ca1.low.S.freq + ggtitle("Cancer (PCAWG)"))
  dev.off()
}

#########################################################################################################

data(UCSC.HG19.Human.CytoBandIdeogram)

RCircos.Set.Core.Components(
  cyto.info      = UCSC.HG19.Human.CytoBandIdeogram,
  chr.exclude    = NULL,
  tracks.inside  = 5,
  tracks.outside = 0)

params                 <- RCircos.Get.Plot.Parameters()
params$base.per.unit   <- 25000
params$text.size       <- .4
params$char.width      <- 1000
params$max.layers      <- 1
params$sub.tracks      <- 1
params$grid.line.color <- "white"
RCircos.Reset.Plot.Parameters(params)

tiff(file="fig_tuples_circos.tiff", width = 7.5 , height = 7.5, units = "in", res = 600, bg = "white", type = "cairo")
RCircos.Set.Plot.Area()
title("Overlapping Regions")

#   Draw chromosome ideogram
RCircos.Chromosome.Ideogram.Plot()

RCircos.Gene.Connector.Plot(genomic.data = cosmic, 
                            track.num    = 1, 
                            side         = "in")

RCircos.Gene.Name.Plot(gene.data = cosmic, 
                       name.col  = 4, 
                       track.num = 2, 
                       side      = "in")

cgi.tiles              <- cgi.sensRegs.sub[, .(Chrom, Start, End)]
cgi.tiles$End          <- cgi.tiles$End + 1
names(cgi.tiles)       <- c("Chromosome", "chromStart", "chromEnd" )

cgi.tiles$value    <- -10
RCircos.Heatmap.Plot(heatmap.data   = cgi.tiles,
                     data.col       = 4,
                     track.num      = 3,
                     side           = "in",
                     min.value      = -10, 
                     max.value      = 10,
                     inside.pos     = 0.72,
                     outside.pos    = 0.99)


ca1.tiles              <- ca1.sensRegs.sub[, .(Chrom, Start, End)]
ca1.tiles$End          <- ca1.tiles$End + 1
names(ca1.tiles)       <- c("Chromosome", "chromStart", "chromEnd" )

ca1.tiles$value    <- 10
RCircos.Heatmap.Plot(heatmap.data   = ca1.tiles, 
                     data.col       = 4,
                     track.num      = 4,
                     side           = "in",
                     min.value      = -10,
                     max.value      = 10,
                     inside.pos     = 0.44,
                     outside.pos    = 0.71)

legend(x      = 0.6, 
       y      = 2.0, 
       legend = c("Normal (CGI)",
                  "Cancer (PCAWG)"), 
       fill   = c("blue", 
                  "red"),
       cex    = 0.8);

dev.off()

#########################################################################################################
#  Plots basic Cluster Statistics as a function of mutational load ( Total [SNVs]): 
#    - Number Clusters, 
#    - Fraction of SNVs in Clusters, 
#    - Fraction of Clusters in genes, 
#    - Cluster Lengths [bp],
#    - Cluster Sizes [SNVs],
#    - Cluster Densities [SNVs/bp]
#                               
#########################################################################################################

n.cltr      <- ggplot()

if (ca1_flag) {
  n.cltr    <- n.cltr + geom_point( data       = ca1.sid.tbl[N_CLTR>0], 
                                    aes(x      = N_SNV, 
                                        y      = N_CLTR,
                                        colour = "ca1"),
                                    shape      = 1,
                                    size       = 3)
}

if (cgi_flag) {
  n.cltr    <- n.cltr + geom_point( data       = cgi.sid.tbl[N_CLTR>0], 
                                    aes(x      = N_SNV, 
                                        y      = N_CLTR,
                                        colour = "cgi"),
                                    shape      = 1,
                                    size       = 3)
}

if (sim_flag) {
  n.cltr    <- n.cltr + geom_errorbar(data        = sim.tbl[Av_N_CLTR>0],
                                      aes( x      = N_SNV,
                                           ymin   = Av_N_CLTR - Sd_N_CLTR, 
                                           ymax   = Av_N_CLTR + Sd_N_CLTR,
                                           colour = "sim"),
                                      size        = 1
  ) + 
    # geom_line( data       = sim.tbl, 
    #            aes(x      = N_SNV, 
    #                y      = Av_N_CLTR,
    #                colour = "sim"),
    #            size   = 1,
    #            linetype = 2
    # ) +
    geom_smooth(data       = sim.tbl[Av_N_CLTR>0], 
                aes(x      = N_SNV, 
                    y      = Av_N_CLTR,
                    colour = "sim"),
                method     = "lm",
                size       = 1,
                linetype   = 2,
                fullrange  = TRUE
    ) #+
  # stat_smooth_func(data       = sim.tbl, 
  #                  aes(x      = N_SNV, 
  #                      y      = Av_N_CLTR,
  #                      colour = "sim"),
  #                  geom   = "text",
  #                  method = "lm",
  #                  hjust  = 0,
  #                  parse  = TRUE,
  #                  xpos   = 60000,
  #                  ypos   = 500,
  #                  size   = 5.5
  # ) 
}

n.cltr    <- n.cltr + coord_cartesian(xlim      = c(300, 1000000-1),
                                      ylim      = c(1, 10000-1),
                                      expand    = FALSE
) +
  scale_x_log10(breaks       = logbreaks,
                minor_breaks = logminor_breaks,
                labels       = comma
  ) + 
  scale_y_log10(breaks       = logbreaks,
                minor_breaks = logminor_breaks,
                labels       = comma
  ) + 
  scale_colour_manual(name   = element_blank(),
                      values = cols,
                      labels = labs,
                      breaks = nams,
                      drop   = TRUE
  ) +
  theme(text         = element_text(size   = 24, 
                                    colour = "black", 
                                    face   = "bold"),
        axis.text    = element_text(size   = rel(0.75)),
        axis.title.x = element_text(size   = rel(1),
                                    margin = margin(1,0,0,0, unit = "lines")),
        axis.title.y = element_text(size   = rel(1),
                                    margin = margin(0,1,0,0, unit = "lines")),
        axis.text.y  = element_text(angle  = 90, 
                                    hjust  = 0.5),
        plot.margin  = unit(c(1, 1, 1, 1), "lines"), #(up, right, down, left)
        plot.title   = element_text(size   = rel(1), 
                                    hjust  = 0,
                                    margin = margin(0,0,1,0, unit = "lines")),
        legend.title          = element_blank(),
        legend.position       = c(0.99, 0.01),
        legend.justification  = c(1, 0),
        legend.text           = element_text(size = 10),
        legend.box.just       = "right", 
        legend.box.background = element_rect()
  ) + 
  labs( x = "Total Number of SNVs",
        y = "Number of Clusters"
  )

#########################################################################################################

f.cltr      <- ggplot() 

if (ca1_flag) {
  f.cltr <- f.cltr + geom_point( data       = ca1.sid.tbl[f>0],
                                 aes(x      = N_SNV, 
                                     y      = f,
                                     colour = "ca1"),
                                 shape      = 1,  
                                 size       = 3)
}

if (cgi_flag) {
  f.cltr <- f.cltr + geom_point( data       = cgi.sid.tbl[f>0],
                                 aes(x      = N_SNV, 
                                     y      = f,
                                     colour = "cgi"),
                                 shape  = 1,  
                                 size   = 3)
}

if (sim_flag) {
  f.cltr <- f.cltr + geom_line( data       = sim.tbl[Av_f>0], 
                                aes(x      = N_SNV, 
                                    y      = Av_f,
                                    colour = "sim"),
                                size       = 1,
                                linetype   = 2 ) +
    geom_errorbar( data        = sim.tbl[Av_f>0],
                   aes( x      = N_SNV,
                        ymin   = Av_f - Sd_f, 
                        ymax   = Av_f + Sd_f,
                        colour = "sim"),
                   size   = 1) 
}

f.cltr <- f.cltr + 
  coord_cartesian(xlim      = c(300, 1000000-1),
                  ylim      = c(0.002, .99),
                  expand    = FALSE
  ) +
  scale_x_log10(breaks       = logbreaks,
                minor_breaks = logminor_breaks,
                labels       = comma
  ) + 
  scale_y_log10(breaks       = c(0.001, 0.01, 0.1, 1),
                minor_breaks = c(seq(0.001, 0.01, 0.001),
                                 seq(0.02, 0.1, 0.01),
                                 seq(0.2,  1,   0.1)),
                labels       = comma
  ) +
  scale_colour_manual(name   = element_blank(),
                      values = cols,
                      labels = labs,
                      breaks = nams,
                      drop  = TRUE
  ) +
  theme(text         = element_text(size   = 24, 
                                    colour = "black", 
                                    face   = "bold"),
        axis.text    = element_text(size   = rel(0.75)),
        axis.title.x = element_text(size   = rel(1),
                                    margin = margin(1,0,0,0, unit = "lines")),
        axis.title.y = element_text(size   = rel(1),
                                    margin = margin(0,1,0,0, unit = "lines")),
        axis.text.y  = element_text(angle  = 90, 
                                    hjust  = 0.5),
        plot.margin  = unit(c(1, 1, 1, 1), "lines"), #(up, right, down, left)
        plot.title   = element_text(size   = rel(1), 
                                    hjust  = 0,
                                    margin = margin(0,0,1,0, unit = "lines")),
        legend.title          = element_blank(),
        legend.position       = c(0.99, 0.01),
        legend.justification  = c(1, 0),
        legend.text           = element_text(size = 10),
        legend.box.just       = "right", 
        legend.box.background = element_rect()
  ) +
  labs( x = "Total Number of SNVs",
        y = "Fraction of SNVs in Clusters"
  ) 

#########################################################################################################

fgenes.cltr      <- ggplot() 

if (ca1_flag) {
  fgenes.cltr <- fgenes.cltr + geom_point( data       = ca1.sid.tbl[f_genes > 0],
                                           aes(x      = N_SNV, 
                                               y      = f_genes,
                                               colour = "ca1"),
                                           shape  = 1,  
                                           size   = 3)
}

if (cgi_flag) {
  fgenes.cltr <- fgenes.cltr + geom_point( data       = cgi.sid.tbl[f_genes > 0],
                                           aes(x      = N_SNV, 
                                               y      = f_genes,
                                               colour = "cgi"),
                                           shape  = 1,  
                                           size   = 3)
}

if (sim_flag) {
  fgenes.cltr <- fgenes.cltr + geom_line( data       = sim.tbl[Av_f_genes > 0], 
                                          aes(x      = N_SNV, 
                                              y      = Av_f_genes,
                                              colour = "sim"),
                                          size   = 1,
                                          linetype = 2) +
    geom_errorbar( data        = sim.tbl[Av_f_genes > 0],
                   aes( x      = N_SNV,
                        ymin   = Av_f_genes - Sd_f_genes, 
                        ymax   = Av_f_genes + Sd_f_genes,
                        colour = "sim"),
                   size   = 1)
}

fgenes.cltr <- fgenes.cltr + 
  coord_cartesian(xlim      = c(300, 1000000-1),
                  ylim      = c(0.01, 0.99),
                  expand    = FALSE
  ) +
  scale_x_log10(breaks       = logbreaks,
                minor_breaks = logminor_breaks,
                labels       = comma
  ) + 
  scale_y_continuous(breaks       = seq(0, 1.2, 0.2),
                     minor_breaks = seq(0, 1.2, 0.01),
                     labels       = comma
  ) +
  #scale_y_log10(breaks       = c(0.001, 0.005, 0.01, 0.05, 0.1, 0.5, 1),
  #              minor_breaks = c(seq(0.001, 0.01, 0.001),
  #                               seq(0.02, 0.1, 0.01),
  #                               seq(0.2,  1,   0.1)),
  #              labels       = comma
  #) +
  scale_colour_manual(name   = element_blank(),
                      values = cols,
                      labels = labs,
                      breaks = nams,
                      drop   = TRUE
  ) + 
  theme(text         = element_text(size   = 24, 
                                    colour = "black", 
                                    face   = "bold"),
        axis.text    = element_text(size   = rel(0.75)),
        axis.title.x = element_text(size   = rel(1),
                                    margin = margin(1,0,0,0, unit = "lines")),
        axis.title.y = element_text(size   = rel(1),
                                    margin = margin(0,1,0,0, unit = "lines")),
        axis.text.y  = element_text(angle  = 90, 
                                    hjust  = 0.5),
        plot.margin  = unit(c(1, 1, 1, 1), "lines"), #(up, right, down, left)
        plot.title   = element_text(size   = rel(1), 
                                    hjust  = 0,
                                    margin = margin(0,0,1,0, unit = "lines")),
        legend.title          = element_blank(),
        legend.position       = c(0.99, 0.99),
        legend.justification  = c(1, 1),
        legend.text           = element_text(size = 10),
        legend.box.just       = "right", 
        legend.box.background = element_rect()
  ) +
  labs( x = "Total Number of SNVs",
        y = "Fraction of cluster SNVs in genes"
  ) 

#########################################################################################################

AvLen.cltr      <- ggplot()

if (ca1_flag) {
  AvLen.cltr <- AvLen.cltr + geom_point( data       = ca1.sid.tbl, 
                                         aes(x      = N_SNV, 
                                             y      = Av.CLTR_Length,
                                             colour = "ca1"),
                                         shape      = 1,  
                                         size       = 3) +
    geom_errorbar( data       = ca1.sid.tbl,
                   aes(x      = N_SNV,
                       ymin   = Av.CLTR_Length - Err.CLTR_Length,
                       ymax   = Av.CLTR_Length + Err.CLTR_Length,
                       colour = "ca1"),
                   size   = .5)
}


if (cgi_flag) {
  AvLen.cltr <- AvLen.cltr  + geom_point( data       = cgi.sid.tbl, 
                                          aes(x      = N_SNV, 
                                              y      = Av.CLTR_Length,
                                              colour = "cgi"),
                                          shape      = 1,  
                                          size       = 3) +
    geom_errorbar( data       = cgi.sid.tbl,
                   aes(x      = N_SNV,
                       ymin   = Av.CLTR_Length - Err.CLTR_Length,
                       ymax   = Av.CLTR_Length + Err.CLTR_Length,
                       colour = "cgi"),
                   size       = .5)
}

if (sim_flag) {
  AvLen.cltr <- AvLen.cltr  + geom_line( data       = sim.tbl, 
                                         aes(x      = N_SNV, 
                                             y      = Av_CL,
                                             colour = "sim"),
                                         size       = 1,
                                         linetype   = 2) +
    geom_errorbar( data        = sim.tbl,
                   aes( x      = N_SNV,
                        ymin   = Av_CL - Sd_CL, 
                        ymax   = Av_CL + Sd_CL,
                        colour = "sim"),
                   size   = 1) 
}

AvLen.cltr <- AvLen.cltr  + 
  coord_cartesian(xlim      = c(300, 1000000-1),
                  ylim      = c(5, 500000),
                  expand    = FALSE
  ) +
  scale_x_log10(breaks       = logbreaks,
                minor_breaks = logminor_breaks,
                labels       = comma
  ) + 
  scale_y_log10(breaks       = c(10, 1000, 100000),
                minor_breaks = logminor_breaks,
                labels       = comma
  ) +
  scale_colour_manual(name   = element_blank(),
                      values = cols,
                      labels = labs,
                      breaks = nams,
                      drop   = TRUE
  ) + 
  theme(text         = element_text(size   = 24, 
                                    colour = "black", 
                                    face   = "bold"),
        axis.text    = element_text(size   = rel(0.75)),
        axis.title.x = element_text(size   = rel(1),
                                    margin = margin(1,0,0,0, unit = "lines")),
        axis.title.y = element_text(size   = rel(1),
                                    margin = margin(0,1,0,0, unit = "lines")),
        axis.text.y  = element_text(angle  = 90, 
                                    hjust  = 0.5),
        plot.margin  = unit(c(1, 1, 1, 1), "lines"), #(up, right, down, left)
        plot.title   = element_text(size   = rel(1), 
                                    hjust  = 0,
                                    margin = margin(0,0,1,0, unit = "lines")),
        legend.title          = element_blank(),
        legend.position       = c(0.99, 0.01),
        legend.justification  = c(1, 0),
        legend.text           = element_text(size = 10),
        legend.box.just       = "right", 
        legend.box.background = element_rect()
  ) +
  labs( x = "Total Number of SNVs",
        y = "Cluster Length [bp]"
  ) 

#########################################################################################################

AvSiz.cltr      <- ggplot() 

if (ca1_flag) {
  AvSiz.cltr <- AvSiz.cltr + geom_point( data       = ca1.sid.tbl, 
                                         aes(x      = N_SNV, 
                                             y      = Av.CLTR_Size,
                                             colour = "ca1"),
                                         shape      = 1,  
                                         size       = 3) +
    geom_errorbar( data       = ca1.sid.tbl, 
                   aes(x      = N_SNV, 
                       ymin   = Av.CLTR_Size - Err.CLTR_Size, 
                       ymax   = Av.CLTR_Size + Err.CLTR_Size,
                       colour = "ca1"),
                   size       = .5)
}

if (cgi_flag) {
  AvSiz.cltr <- AvSiz.cltr + geom_point( data       = cgi.sid.tbl, 
                                         aes(x      = N_SNV, 
                                             y      = Av.CLTR_Size,
                                             colour = "cgi"),
                                         shape      = 1,  
                                         size       = 3) +
    geom_errorbar( data       = cgi.sid.tbl, 
                   aes(x      = N_SNV, 
                       ymin   = Av.CLTR_Size - Err.CLTR_Size, 
                       ymax   = Av.CLTR_Size + Err.CLTR_Size,
                       colour = "cgi"),
                   size       = .5) 
}

if (sim_flag) {
  AvSiz.cltr <- AvSiz.cltr + geom_line( data       = sim.tbl, 
                                        aes(x      = N_SNV, 
                                            y      = Av_CS,
                                            colour = "sim"),
                                        size       = 1,
                                        linetype   = 2) +
    geom_errorbar( data        = sim.tbl,
                   aes( x      = N_SNV,
                        ymin   = Av_CS - Sd_CS, 
                        ymax   = Av_CS + Sd_CS,
                        colour = "sim"),
                   size        = 1)
}

AvSiz.cltr <- AvSiz.cltr + 
  coord_cartesian(xlim      = c(300, 1000000-1),
                  ylim      = c(2.5, 300),
                  expand    = FALSE
  ) +
  scale_x_log10(breaks       = logbreaks,
                minor_breaks = logminor_breaks,
                labels       = comma
  ) + 
  scale_y_log10(breaks       = c(1, 5, 10, 50, 100, 500),
                minor_breaks = logminor_breaks,
                labels       = comma
  ) +
  scale_colour_manual(name   = element_blank(),
                      values = cols,
                      labels = labs,
                      breaks = nams,
                      drop   = TRUE) + 
  theme(text         = element_text(size   = 24, 
                                    colour = "black", 
                                    face   = "bold"),
        axis.text    = element_text(size   = rel(0.75)),
        axis.title.x = element_text(size   = rel(1),
                                    margin = margin(1,0,0,0, unit = "lines")),
        axis.title.y = element_text(size   = rel(1),
                                    margin = margin(0,1,0,0, unit = "lines")),
        axis.text.y  = element_text(angle  = 90, 
                                    hjust  = 0.5),
        plot.margin  = unit(c(1, 1, 1, 1), "lines"), #(up, right, down, left)
        plot.title   = element_text(size   = rel(1), 
                                    hjust  = 0,
                                    margin = margin(0,0,1,0, unit = "lines")),
        legend.title          = element_blank(),
        legend.position       = c(0.01, .99),
        legend.justification  = c(0, 1),
        legend.text           = element_text(size = 10),
        legend.box.just       = "right", 
        legend.box.background = element_rect()
  ) + 
  labs( x = "Total Number of SNVs",
        y = "Cluster Size [SNV]"
  ) 

#########################################################################################################

AvDen.cltr      <- ggplot()

if (ca1_flag) {
  AvDen.cltr <- AvDen.cltr + geom_point( data   = ca1.sid.tbl,
                                         aes(x  = N_SNV, 
                                             y  = (Av.CLTR_Density),
                                             colour = "ca1"),
                                         shape  = 1,  
                                         size   = 3) +
    geom_errorbar( data     = ca1.sid.tbl,
                   aes(x    = N_SNV, 
                       ymin = (Av.CLTR_Density - Err.CLTR_Density), 
                       ymax = (Av.CLTR_Density + Err.CLTR_Density),
                       colour = "ca1"),
                   size   = .5)
}

if (cgi_flag) {
  AvDen.cltr <- AvDen.cltr + geom_point( data   = cgi.sid.tbl,
                                         aes(x  = N_SNV, 
                                             y  = (Av.CLTR_Density),
                                             colour = "cgi"),
                                         shape  = 1,  
                                         size   = 3) +
    geom_errorbar( data     = cgi.sid.tbl,
                   aes(x    = N_SNV, 
                       ymin = (Av.CLTR_Density - Err.CLTR_Density), 
                       ymax = (Av.CLTR_Density + Err.CLTR_Density),
                       colour = "cgi"),
                   size   = .5)
}

if (sim_flag) {
  AvDen.cltr <- AvDen.cltr + geom_line( data   = sim.tbl, 
                                        aes(x  = N_SNV, 
                                            y  = (Av_CD),
                                            colour = "sim"),
                                        size   = 1,
                                        linetype = 2) +
    geom_errorbar( data   = sim.tbl,
                   aes( x    = N_SNV,
                        ymin = (Av_CD - Sd_CD), 
                        ymax = (Av_CD + Sd_CD),
                        colour = "sim"),
                   size   = 1)
}

AvDen.cltr <- AvDen.cltr + 
  coord_cartesian(xlim      = c(300, 1000000-1),
                  ylim      = c(3e-5, 0.99),
                  expand    = FALSE
  ) +
  scale_x_log10(breaks       = logbreaks,
                minor_breaks = logminor_breaks,
                labels       = comma
  ) + 
  scale_y_log10(breaks       = c(1e-4, 1e-2, 1),
                minor_breaks = c(seq(1e-5, 1e-4, 1e-5),
                                 seq(2e-4, 1e-3, 1e-4),
                                 seq(2e-3, 1e-2, 1e-3),
                                 seq(2e-2, 1e-1, 1e-2),
                                 seq(2e-1, 1,    1e-1)),
                labels       = comma
  ) +
  scale_colour_manual(name   = element_blank(),
                      values = cols,
                      labels = labs,
                      breaks = nams,
                      drop   = TRUE
  ) + 
  theme(text         = element_text(size   = 24, 
                                    colour = "black", 
                                    face   = "bold"),
        axis.text    = element_text(size   = rel(0.75)),
        axis.title.x = element_text(size   = rel(1),
                                    margin = margin(1,0,0,0, unit = "lines")),
        axis.title.y = element_text(size   = rel(1),
                                    margin = margin(0,1,0,0, unit = "lines")),
        axis.text.y  = element_text(angle  = 90, 
                                    hjust  = 0.5),
        plot.margin  = unit(c(1, 1, 1, 1), "lines"), #(up, right, down, left)
        plot.title   = element_text(size   = rel(1), 
                                    hjust  = 0,
                                    margin = margin(0,0,1,0, unit = "lines")),
        legend.title          = element_blank(),
        legend.position       = c(0.99, 0.99),
        legend.justification  = c(1, 1),
        legend.text           = element_text(size = 10),
        legend.box.just       = "right", 
        legend.box.background = element_rect()
  ) +
  labs( x = "Total Number of SNVs",
        y = "Cluster Density [SNV/bp]"
  ) 

#########################################################################################################

S_V_N.cltr      <- ggplot()

if (ca1_flag) {
  S_V_N.cltr <- S_V_N.cltr + geom_point( data   = ca1.cltr.tbl,
                                         aes(x  = N_Mutations-.1, 
                                             y  = Length,
                                             colour = "ca1"),
                                         shape  = 20,  
                                         size   = .5) 
}

if (cgi_flag) {
  S_V_N.cltr <- S_V_N.cltr + geom_point( data   = cgi.cltr.tbl,
                                         aes(x  = N_Mutations+.1, 
                                             y  = Length,
                                             colour = "cgi"),
                                         shape  = 20,  
                                         size   = .5)
}

if (sim_flag) {
  S_V_N.cltr <- S_V_N.cltr + geom_point( data   = sim.cltr.tbl,
                                         aes(x  = N_Mutations, 
                                             y  = Length,
                                             colour = "sim"),
                                         shape  = 20,  
                                         size   = .5) 
}

xx      = seq(3, 10000, 100)

S_V_N.cltr <- S_V_N.cltr +
  geom_line( aes(x      = xx,
                 y      = (xx-1)*25000,
                 colour = "sim"),
             size   = 1,
             linetype = 2) +
  geom_line( aes(x      = xx,
                 y      = xx,
                 colour = "sim"),
             size   = 1,
             linetype = 2) +
  coord_cartesian(xlim      = c(2, 10000 - 1),
                  ylim      = c(2, 10000000 - 1),
                  expand    = FALSE
  ) +
  scale_x_log10(breaks       = logbreaks,
                minor_breaks = logminor_breaks,
                labels       = comma
  ) + 
  scale_y_log10(breaks       = c(10, 1000, 100000, 10000000),
                minor_breaks = logminor_breaks,
                labels       = comma
  ) +
  scale_colour_manual(name   = element_blank(),
                      values = cols,
                      labels = labs,
                      breaks = nams,
                      drop   = TRUE
  ) + 
  theme(text         = element_text(size   = 24, 
                                    colour = "black", 
                                    face   = "bold"),
        axis.text    = element_text(size   = rel(0.75)),
        axis.title.x = element_text(size   = rel(1),
                                    margin = margin(1,0,0,0, unit = "lines")),
        axis.title.y = element_text(size   = rel(1),
                                    margin = margin(0,1,0,0, unit = "lines")),
        axis.text.y  = element_text(angle  = 90, 
                                    hjust  = 0.5),
        plot.margin  = unit(c(1, 1, 1, 1), "lines"), #(up, right, down, left)
        plot.title   = element_text(size   = rel(1), 
                                    hjust  = 0,
                                    margin = margin(0,0,1,0, unit = "lines")),
        legend.title          = element_blank(),
        legend.position       = c(0.99, 0.01),
        legend.justification  = c(1, 0),
        legend.text           = element_text(size = 10),
        legend.box.just       = "right", 
        legend.box.background = element_rect()
  ) +
  labs( x = "Cluster Size [SNV]",
        y = "Cluster Length [bp]"
  ) 

#########################################################################################################

lS_V_N.cltr      <- ggplot()

if (ca1_flag) {
  lS_V_N.cltr <- lS_V_N.cltr + geom_point( data   = ca1.sid.tbl,
                                           aes(x  = Av.CLTR_Size, 
                                               y  = Av.CLTR_Length,
                                               colour = "ca1"),
                                           shape  = 20,  
                                           size   = .5) +
    geom_errorbar( data     = ca1.sid.tbl,
                   aes(x    = Av.CLTR_Size, 
                       ymin = (Av.CLTR_Length - Err.CLTR_Length), 
                       ymax = (Av.CLTR_Length + Err.CLTR_Length),
                       colour = "ca1"),
                   size   = .5) +
    geom_errorbarh( data     = ca1.sid.tbl,
                    aes(x    = Av.CLTR_Size,
                        y    = Av.CLTR_Length, 
                        xmin = (Av.CLTR_Size - Err.CLTR_Size), 
                        xmax = (Av.CLTR_Size + Err.CLTR_Size),
                        colour = "ca1"),
                    size   = .5) 
}

if (cgi_flag) {
  lS_V_N.cltr <- lS_V_N.cltr + geom_point( data   = cgi.sid.tbl,
                                           aes(x  = Av.CLTR_Size, 
                                               y  = Av.CLTR_Length,
                                               colour = "cgi"),
                                           shape  = 20,  
                                           size   = .5) +
    geom_errorbar( data     = cgi.sid.tbl,
                   aes(x    = Av.CLTR_Size, 
                       ymin = (Av.CLTR_Length - Err.CLTR_Length), 
                       ymax = (Av.CLTR_Length + Err.CLTR_Length),
                       colour = "cgi"),
                   size   = .5) +
    geom_errorbarh( data     = cgi.sid.tbl,
                    aes(x    = Av.CLTR_Size,
                        y  = Av.CLTR_Length, 
                        xmin = (Av.CLTR_Size - Err.CLTR_Size), 
                        xmax = (Av.CLTR_Size + Err.CLTR_Size),
                        colour = "cgi"),
                    size   = .5)
}

if (sim_flag) {
  lS_V_N.cltr <- lS_V_N.cltr + geom_point( data   = sim.sid.tbl,
                                           aes(x  = Av.CLTR_Size, 
                                               y  = Av.CLTR_Length,
                                               colour = "sim"),
                                           shape  = 20,  
                                           size   = .5) +
    geom_errorbar( data     = sim.sid.tbl,
                   aes(x    = Av.CLTR_Size, 
                       ymin = (Av.CLTR_Length - Err.CLTR_Length), 
                       ymax = (Av.CLTR_Length + Err.CLTR_Length),
                       colour = "sim"),
                   size   = .5) +
    geom_errorbarh( data     = sim.sid.tbl,
                    aes(x    = Av.CLTR_Size, 
                        y  = Av.CLTR_Length, 
                        xmin = (Av.CLTR_Size - Err.CLTR_Size), 
                        xmax = (Av.CLTR_Size + Err.CLTR_Size),
                        colour = "sim"),
                    size   = .5)
}

xx      = seq(3, 10000, 100)

lS_V_N.cltr <- lS_V_N.cltr +
  geom_line( aes(x      = xx,
                 y      = (xx - 1)*25000,
                 colour = "sim"),
             size   = 1,
             linetype = 2) +
  geom_line( aes(x      = xx,
                 y      = xx,
                 colour = "sim"),
             size   = 1,
             linetype = 2) +
  coord_cartesian(xlim      = c(2, 300),
                  ylim      = c(7, 400000),
                  expand    = FALSE
  ) +
  scale_x_log10(breaks       = logbreaks,
                minor_breaks = logminor_breaks,
                labels       = comma
  ) + 
  scale_y_log10(breaks       = c(10, 1000, 100000, 10000000),
                minor_breaks = logminor_breaks,
                labels       = comma
  ) +
  scale_colour_manual(name   = element_blank(),
                      values = cols,
                      labels = labs,
                      breaks = nams,
                      drop   = TRUE
  ) + 
  theme(text         = element_text(size   = 24, 
                                    colour = "black", 
                                    face   = "bold"),
        axis.text    = element_text(size   = rel(0.75)),
        axis.title.x = element_text(size   = rel(1),
                                    margin = margin(1,0,0,0, unit = "lines")),
        axis.title.y = element_text(size   = rel(1),
                                    margin = margin(0,1,0,0, unit = "lines")),
        axis.text.y  = element_text(angle  = 90, 
                                    hjust  = 0.5),
        plot.margin  = unit(c(1, 1, 1, 1), "lines"), #(up, right, down, left)
        plot.title   = element_text(size   = rel(1), 
                                    hjust  = 0,
                                    margin = margin(0,0,1,0, unit = "lines")),
        legend.title          = element_blank(),
        legend.position       = c(0.99, 0.01),
        legend.justification  = c(1, 0),
        legend.text           = element_text(size = 10),
        legend.box.just       = "right", 
        legend.box.background = element_rect()
  ) +
  labs( x = "Mean Cluster Size [SNV]",
        y = "Mean Cluster Length [bp]"
  )


fileout <- paste("fig_clusters_A.tiff", sep="")
#tiff(file=fileout, width = 4500 , height = 4500, units = "px", res = 600, bg = "white")
tiff(file=fileout, width = 7.5 , height = 7.5, units = "in", type="cairo", res = 300)
print(n.cltr)
dev.off()

fileout <- paste("fig_clusters_B.tiff", sep="")
#tiff(file=fileout, width = 4500 , height = 4500, units = "px", res = 600, bg = "white")
tiff(file=fileout, width = 7.5 , height = 7.5, units = "in", type="cairo", res = 300)
print(f.cltr)
dev.off()

fileout <- paste("fig_clusters_C.tiff", sep="")
#tiff(file=fileout, width = 4500 , height = 4500, units = "px", res = 600, bg = "white")
tiff(file=fileout, width = 7.5 , height = 7.5, units = "in", type="cairo", res = 300)
print(fgenes.cltr)
dev.off()

fileout <- paste("fig_clusters_D.tiff", sep="")
#tiff(file=fileout, width = 4500 , height = 4500, units = "px", res = 600, bg = "white")
tiff(file=fileout, width = 7.5 , height = 7.5, units = "in", type="cairo", res = 300)
print(AvLen.cltr)
dev.off()

fileout <- paste("fig_clusters_E.tiff", sep="")
#tiff(file=fileout, width = 4500 , height = 4500, units = "px", res = 600, bg = "white")
tiff(file=fileout, width = 7.5 , height = 7.5, units = "in", type="cairo", res = 300)
print(AvSiz.cltr)
dev.off()

fileout <- paste("fig_clusters_F.tiff", sep="")
#tiff(file=fileout, width = 4500 , height = 4500, units = "px", res = 600, bg = "white")
tiff(file=fileout, width = 7.5 , height = 7.5, units = "in", type="cairo", res = 300)
print(AvDen.cltr)
dev.off()

fileout <- paste("fig_clusters_G.tiff", sep="")
#tiff(file=fileout, width = 4500 , height = 4500, units = "px", res = 600, bg = "white")
tiff(file=fileout, width = 7.5 , height = 7.5, units = "in", type="cairo", res = 300)
print(S_V_N.cltr) 
dev.off()

fileout <- paste("fig_clusters_H.tiff", sep="")
#tiff(file=fileout, width = 4500 , height = 4500, units = "px", res = 600, bg = "white")
tiff(file=fileout, width = 7.5 , height = 7.5, units = "in", type="cairo", res = 300)
print(lS_V_N.cltr)
dev.off()
